compiler/rustc_const_eval/src/check_consts/qualifs.rs

```diff
@@ -192,16 +183,32 @@ impl Qualif for NeedsNonConstDrop {
             return false;
         }
 
-        // FIXME(const_trait_impl): Reimplement const drop checking.
-        NeedsDrop::in_any_value_of_ty(cx, ty)
+        if cx.tcx.features().const_trait_impl() {
```

Comment:
That seems odd to have this analysis depend on whether the *current* crate has the feature enabled?

Reply:
Should we gate this some other way?

Reply:
What I would expect is that in `check_consts/check.rs`, we have something like
- if not `NeedsDrop`, then accept
- otherwise, fire `ops::LiveDrop`, telling it about the `NeedsNonConstDrop` qualif
  - if that qualif is `true`, make it `Status::Forbidden`
  - otherwise, make it `Status::Unstable` with a suitable feature gate (IMO this should be a separate gate from `const_trait_impl`)

Reply:
I believe I implemented what you asked for in the last commit. First we check `NeedsDrop`, then we check `NeedsNonConstDrop` and pass that to `LiveDrop` to differentiate the status.

I also needed to change the analogous logic for checking `const_precise_live_drops`, which happens in a post-drop-elaboration MIR lint.

Reply:
also: https://github.com/rust-lang/rust/pull/132329#discussion_r1838777932